File: RUBiS-1.4.3/Servlets/edu/rice/rubis/servlets/ViewBidHistory.java
package edu.rice.rubis.servlets;

import java.io.IOException;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

/** This servlets displays the list of bids regarding an item.
 * It must be called this way :
 * <pre>
 * http://..../ViewUserInfo?itemId=xx where xx is the id of the item
 * /<pre>
 */

public class ViewBidHistory extends RubisHttpServlet
{


  public int getPoolSize()
  {
    return Config.ViewBidHistoryPoolSize;
  }

/**
 * Close both statement and connection to the database.
 */
  private void closeConnection(PreparedStatement stmt, Connection conn)
  {
    try
    {
      if (stmt != null)
        stmt.close(); // close statement
      if (conn != null)
        releaseConnection(conn);
    }
    catch (Exception ignore)
    {
    }
  }

  /** List the bids corresponding to an item */
  private boolean listBids(Integer itemId, PreparedStatement stmt, Connection conn, ServletPrinter sp)
  {
    float bid;
    int userId;
    String bidderName, date;
    ResultSet rs = null;

    // Get the list of the user's last bids
    try
    {
      stmt =
        conn.prepareStatement(
          "SELECT * FROM bids WHERE item_id=? ORDER BY date DESC");
      stmt.setInt(1, itemId.intValue());
      rs = stmt.executeQuery();
      if (!rs.first())
      {
        sp.printHTML(
          "<h3>There is no bid corresponding to this item.</h3><br>");
        closeConnection(stmt, conn);
        return false;
      }
    }
    catch (SQLException e)
    {
      sp.printHTML("Exception getting bids list: " + e + "<br>");
      closeConnection(stmt, conn);
      return false;
    }

    sp.printBidHistoryHeader();
    try
    {
      do
      {
        // Get the bids
        date = rs.getString("date");
        bid = rs.getFloat("bid");
        userId = rs.getInt("user_id");

        ResultSet urs = null;
        try
        {
          stmt = conn.prepareStatement("SELECT nickname FROM users WHERE id=?");
          stmt.setInt(1, userId);
          urs = stmt.executeQuery();
          if (!urs.first())
          {
            sp.printHTML("This user does not exist in the database.<br>");
            closeConnection(stmt, conn);
            return false;
          }
          bidderName = urs.getString("nickname");
        }
        catch (SQLException e)
        {
          sp.printHTML("Couldn't get bidder name: " + e + "<br>");
          closeConnection(stmt, conn);
          return false;
        }
        sp.printBidHistory(userId, bidderName, bid, date);
      }
      while (rs.next());
    }
    catch (SQLException e)
    {
      sp.printHTML("Exception getting bid: " + e + "<br>");
      closeConnection(stmt, conn);
      return false;
    }
    sp.printBidHistoryFooter();
    return true;
  }

  public void doGet(HttpServletRequest request, HttpServletResponse response)
    throws IOException, ServletException
  {
    doPost(request, response);
  }

  public void doPost(HttpServletRequest request, HttpServletResponse response)
    throws IOException, ServletException
  {
    String value = request.getParameter("itemId");
    Integer itemId;
    String itemName;
    ResultSet rs = null;
    ServletPrinter sp = null;
    PreparedStatement stmt = null;
    Connection conn = null;

    sp = new ServletPrinter(response, "ViewBidHistory");

    if ((value == null) || (value.equals("")))
    {
      sp.printHTMLheader("RUBiS ERROR: View bids history");
      sp.printHTML("<h3>You must provide an item identifier !<br></h3>");
      sp.printHTMLfooter();
      return;
    }
    else
      itemId = new Integer(value);
    if (itemId.intValue() == -1)
      sp.printHTML("ItemId is -1: this item does not exist.<br>");

    sp.printHTMLheader("RUBiS: Bid history");

    // get the item
    try
    {
      conn = getConnection();
      stmt = conn.prepareStatement("SELECT name FROM items WHERE id=?");
      stmt.setInt(1, itemId.intValue());
      rs = stmt.executeQuery();
    }
    catch (Exception e)
    {
      sp.printHTML("Failed to execute Query for item in table items: " + e);
      closeConnection(stmt, conn);
      return;
    }
    /**
    try
    {
      if (!rs.first())
      {
        stmt.close();
        stmt = conn.prepareStatement("SELECT name FROM old_items WHERE id=?");
        stmt.setInt(1, itemId.intValue());
        rs = stmt.executeQuery();
      }
    }
    catch (Exception e)
    {
      sp.printHTML("Failed to execute Query for item in table old_items: " + e);
      closeConnection(stmt, conn);
      return;
    }
    */
    try
    {
      if (!rs.first())
      {
        sp.printHTML("<h2>This item does not exist!</h2>");
        closeConnection(stmt, conn);
        return;
      }
      itemName = rs.getString("name");
      sp.printHTML(
        "<center><h3>Bid History for " + itemName + "<br></h3></center>");
    }
    catch (Exception e)
    {
      sp.printHTML("This item does not exist (got exception: " + e + ")<br>");
      sp.printHTMLfooter();
      closeConnection(stmt, conn);
      return;
    }

    boolean connAlive = listBids(itemId, stmt, conn, sp);
    // connAlive means we must close it. Otherwise we must NOT do a
    // double free
    if (connAlive) {
        closeConnection(stmt, conn);
    }
    sp.printHTMLfooter();
  }

  /**
  * Clean up the connection pool.
  */
  public void destroy()
  {
    super.destroy();
  }

}

File: RUBiS-1.4.3/PHP/ViewBidHistory.php
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <body>
    <?php
    $scriptName = "ViewBidHistory.php";
    include("PHPprinter.php");
    $startTime = getMicroTime();
    
    $itemId = $HTTP_POST_VARS['itemId'];
    if ($itemId == null)
    {
      $itemId = $HTTP_GET_VARS['itemId'];
      if ($itemId == null)
      {
         printError($scriptName, $startTime, "Bid history", "You must provide an item identifier!<br>");
         exit();
      }
    }
      
    getDatabaseLink($link);
    beginRO($link);

function getItemBids($link, $itemId)
{
    txcache_invaltag("bids", "item_id", $itemId);
    $bidsListResult = sql_query("SELECT * FROM bids WHERE item_id=$itemId ORDER BY date DESC", $link) or die("ERROR: Bids list query failed");
    $bids = array();
    
    while ($bidsListRow = sql_fetch_array($bidsListResult)) {
        $bids[] = $bidsListRow;
    }

    sql_free_result($bidsListResult);
    return $bids;
}

function viewbidhistory($link, $itemId) {
    // Get the item name
    $itemNameRow = getAnyItem($link, $itemId);
    if (!$itemNameRow)
    {
      die("<h3>ERROR: Sorry, but this item does not exist.</h3><br>\n");
    }
    $itemName = $itemNameRow["name"];

    
    // Get the list of bids for this item
    $bids = wrap(true, 'getItemBids', $link, $itemId);
    if (count($bids) == 0)
      print ("<h2>There is no bid for $itemName. </h2><br>");
    else
      print ("<h2><center>Bid history for $itemName</center></h2><br>");

    printHTMLheader("RUBiS: Bid history for $itemName.");
    print("<TABLE border=\"1\" summary=\"List of bids\">\n".
                "<THEAD>\n".
                "<TR><TH>User ID<TH>Bid amount<TH>Date of bid\n".
                "<TBODY>\n");

    foreach ($bids as $bidsListRow)
    {
    	$bidAmount = $bidsListRow["bid"];
    	$bidDate = $bidsListRow["date"];
    	$userId = $bidsListRow["user_id"];
	// Get the bidder nickname	
    	if ($userId != 0)
	{
          $userNameRow = getUser($link, $userId);
	  $nickname = $userNameRow["nickname"];
    	}
    	else
	  {
	    print("Cannot lookup the user!<br>");
	    printHTMLfooter($scriptName, $startTime);
	    exit();
	  }
        print("<TR><TD><a href=\"/PHP/ViewUserInfo.php?userId=".$userId."\">$nickname</a>"
		  ."<TD>".$bidAmount."<TD>".$bidDate."\n");
    }
    print("</TABLE>\n");

    sql_free_result($bidsListResult);
}

wrap(true, 'viewbidhistory', $link, $itemId);

    commit($link);
    sql_close($link);
    
    printHTMLfooter($scriptName, $startTime);
    ?>
  </body>
</html>
